presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,261 @@
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdown();
+    }
+
+    private void startDiscovery()
+    {
+        long startDelay = getStartDelay();
+
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                long interval = (long) discoveryInterval.convertTo(SECONDS).getValue();
+                SECONDS.sleep(ThreadLocalRandom.current().nextLong(1, interval));
+
+                log.info("Running shard organizer...");
+                discoverShardsAndSubmitJobs();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error running shard organizer");
+            }
+            finally {
+                shardOrganizerDao.updateShardOrganizerLastRunTime(currentNodeIdentifier);
+            }
+        }, startDelay, discoveryInterval.toMillis(), MILLISECONDS);
+    }
+
+    private long getStartDelay()
+    {
+        Long secondsAgo = shardOrganizerDao.getShardOrganizerLastStartTime(currentNodeIdentifier);
+        if (secondsAgo == null) {
+            shardOrganizerDao.insertNode(currentNodeIdentifier);
+            secondsAgo = shardOrganizerDao.getShardOrganizerLastStartTime(currentNodeIdentifier);
+        }
+        return Math.max(discoveryInterval.toMillis() - MILLISECONDS.convert(secondsAgo, SECONDS), 0);
+    }
+
+    @VisibleForTesting
+    void discoverShardsAndSubmitJobs()
+    {
+        Collection<Long> enabledTableIds = metadataDao.listOrganizationEnabledTables();
+        Set<ShardMetadata> allShards = shardManager.getNodeShards(currentNodeIdentifier);
+        ListMultimap<Long, ShardMetadata> tableShards = Multimaps.index(allShards, ShardMetadata::getTableId);
+
+        for (Map.Entry<Long, Collection<ShardMetadata>> entry : tableShards.asMap().entrySet()) {
+            long tableId = entry.getKey();
+
+            // If not enabled or not sorted, nothing to organize, skip the table
+            if (!enabledTableIds.contains(tableId) || metadataDao.listSortColumns(tableId).isEmpty()) {
+                continue;
+            }
+
+            Table tableInfo = metadataDao.getTableInformation(tableId);
+            Set<ShardMetadata> filteredShards = entry.getValue().stream()
+                    .filter(shard -> !organizer.inProgress(shard.getShardUuid()))
+                    .collect(Collectors.toSet());
+
+            Collection<ShardIndexInfo> indexInfos = toShardIndexInfo(dbi, metadataDao, tableInfo, filteredShards, true);
+            Set<OrganizationSet> organizationSets = createOrganizationSets(tableInfo, indexInfos);
+
+            log.info("Table ID: %s Shards selected: %s Organization sets created: %s", tableId, filteredShards.size(), organizationSets.size());
```

Comment:
I don't like this logging format because it makes the fields run together. How about
```java
log.info("Created %s organization set(s) from %s shard(s) for table ID %s", organizationSets.size(), filteredShards.size(), tableId);